node/cmd/guardiand/adminserver.go

```diff
@@ -539,3 +564,106 @@ func (s *nodePrivilegedService) PurgePythNetVaas(ctx context.Context, req *nodev
+	v, err := vaa.Unmarshal(req.Vaa)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unmarshal VAA: %w", err)
+	}
+
+	if req.NewGuardianSetIndex <= v.GuardianSetIndex {
+		return nil, errors.New("new guardian set index must be higher than provided VAA")
+	}
+
+	if s.evmConnector == nil {
+		return nil, errors.New("the node needs to have an Ethereum connection configured to sign existing VAAs")
+	}
+
+	var gs *common.GuardianSet
+	if cachedGs, exists := s.gsCache.Load(v.GuardianSetIndex); exists {
+		gs = cachedGs.(*common.GuardianSet)
+	} else {
+		evmGs, err := s.evmConnector.GetGuardianSet(ctx, v.GuardianSetIndex)
+		if err != nil {
+			return nil, fmt.Errorf("failed to load guardian set [%d]: %w", v.GuardianSetIndex, err)
+		}
+		gs = &common.GuardianSet{
+			Keys:  evmGs.Keys,
+			Index: v.GuardianSetIndex,
+		}
+		s.gsCache.Store(v.GuardianSetIndex, gs)
+	}
+
+	if slices.Index(gs.Keys, s.guardianAddress) != -1 {
+		return nil, fmt.Errorf("local guardian is already on the old set")
+	}
+
+	// Verify VAA
+	err = v.Verify(gs.Keys)
+	if err != nil {
+		return nil, fmt.Errorf("failed to verify existing VAA: %w", err)
+	}
+
+	if len(req.NewGuardianSet) > 255 {
+		return nil, errors.New("new guardian set has too many guardians")
+	}
+	newGS := make([]ethcommon.Address, len(req.NewGuardianSet))
+	for i, guardianString := range req.NewGuardianSet {
+		guardianAddress := ethcommon.HexToAddress(guardianString)
+		newGS[i] = guardianAddress
+	}
+	// Make sure there are no duplicates
+	newGsLen := len(newGS)
+	if len(slices.Compact(newGS)) != newGsLen {
+		return nil, fmt.Errorf("duplicate guardians in the guardian set")
+	}
+
+	localGuardianIndex := slices.Index(newGS, s.guardianAddress)
+	if localGuardianIndex == -1 {
+		return nil, fmt.Errorf("local guardian is not a member of the new guardian set")
+	}
+
+	newVAA := &vaa.VAA{
+		Version: v.Version,
+		// Set the new guardian set index
+		GuardianSetIndex: req.NewGuardianSetIndex,
+		// Signatures will be repopulated
+		Signatures:       nil,
+		Timestamp:        v.Timestamp,
+		Nonce:            v.Nonce,
+		Sequence:         v.Sequence,
+		ConsistencyLevel: v.ConsistencyLevel,
+		EmitterChain:     v.EmitterChain,
+		EmitterAddress:   v.EmitterAddress,
+		Payload:          v.Payload,
+	}
+
+	// Copy original VAA signatures
+	for _, sig := range v.Signatures {
+		signerAddress := gs.Keys[sig.Index]
+		newIndex := slices.Index(newGS, signerAddress)
+		// Guardian is not part of the new set
+		if newIndex == -1 {
+			continue
+		}
+		newVAA.Signatures = append(newVAA.Signatures, &vaa.Signature{
+			Index:     uint8(newIndex),
+			Signature: sig.Signature,
+		})
+	}
+
+	// Add our own signature only if the new guardian set would reach quorum
+	if vaa.CalculateQuorum(len(newGS)) > len(newVAA.Signatures)+1 {
+		return nil, errors.New("cannot reach quorum on new guardian set with the local signature")
+	}
+
+	// Add local signature
+	newVAA.AddSignature(s.gk, uint8(localGuardianIndex))
```

Comment:
If I understand correctly, many contracts expect guardian indexes to be monotonically increasing when verifying signatures so adding this signature here could cause signature verification to fail if `localGuardianIndex` is some low number like 0 or 5.

I think it would be better if those contracts used a bitset to keep track of signers rather than requiring strictly increasing index values but sadly that ship has sailed.